src/search.py

```diff
@@ -0,0 +1,148 @@
+        for result in search_results:
+            set_name = result["set_name"]
+            data_keywords_df = data_keywords_df.merge(result["df"][["id", f"_score_{set_name}"]], on="id", how="left")
+        keyword_set_names = [result["set_name"] for result in search_results]
+        data_keywords_df = (
+            data_keywords_df.fillna(
+                {f"_score_{set_name}": 0 for set_name in self.config["search_recipe"]["keyword_sets"]}
+            )
+            .assign(_score=lambda df: df[[f"_score_{set_name}" for set_name in keyword_set_names]].sum(axis=1))
+            # Normalise score
+            .assign(_score=lambda df: (df._score - df._score.min()) / (df._score.max() - df._score.min()))
+            .sort_values("_score", ascending=False)
+            .reset_index()
+        )
+
+        return data_keywords_df
+
+    def vector_searches(self, scope_statements: list) -> pd.DataFrame:
+        """Perform vector searches"""
+        search_results = []
+        for statement in scope_statements:
+            search_results.append(
+                self.Dataset.vector_search(statement, n_results=self._n_vector_results).assign(statement=statement)
+            )
+
+        search_results = (
+            pd.concat(search_results, ignore_index=True)
+            .groupby("id")
+            .agg(
+                n_counts=("id", "count"),
+                _distance=("_distance", "min"),
+            )
+            .assign(
+                _score=lambda df: 1 - (df._distance - df._distance.min()) / (df._distance.max() - df._distance.min())
+            )
+            .drop(columns="_distance")
+            .reset_index()
+        )
+        return search_results
+
+    def vector_search_results(self, vector_searches_df: pd.DataFrame) -> pd.DataFrame:
+        """Organise vector search results into a single dataframe"""
+        return (
+            self.data_df.query("id in @vector_searches_df.id")
+            .merge(vector_searches_df, how="left", on="id")
+            .sort_values("_score", ascending=False)
+        )
+
+    def final_results(self, data_keywords_df: pd.DataFrame, data_vectors_df: pd.DataFrame) -> pd.DataFrame:
+        """Combine both searches"""
+        final_ids = set(data_keywords_df.id).union(set(data_vectors_df.id))  # noqa
+        final_matches_df = (
+            self.data_df.query("id in @final_ids")
+            .merge(data_keywords_df[["id", "_score"]], how="left", on="id")
+            .rename(columns={"_score": "_score_keywords"})
+            .merge(data_vectors_df[["id", "_score"]], how="left", on="id")
+            .rename(columns={"_score": "_score_vectors"})
+            .fillna({"_score_keywords": 0, "_score_vectors": 0})
+            .assign(_score=lambda df: (df._score_keywords + df._score_vectors) / 2)
+            .sort_values("_score", ascending=False)
+            .reset_index()
+        )
+
+        return final_matches_df
+
+    def do_search(self) -> pd.DataFrame:
```

Comment:
A docstring for this function or the previous one would be nice to make it clearer what the output contains and how that should be interpreted. Either describe how `_score_keywords`, `_score_vectors` and `_score` are calculated in the functions that generate them and then refer to those functions here, or describe them all here.